app/services/oauth_clients/redirect_uri_from_state_service.rb

```diff
@@ -28,30 +28,45 @@
 
 require "rack/oauth2"
 require "uri/http"
+require 'dry/monads'
+require 'dry/monads/do'
 
 module OAuthClients
   class RedirectUriFromStateService
+    include Dry::Monads[:maybe]
+    include Dry::Monads::Do.for(:process)
+
     def initialize(state:, cookies:)
-      @state = state
+      @state = Maybe(state)
       @cookies = cookies
     end
 
     def call
-      redirect_uri = oauth_state_cookie
-
-      if redirect_uri.present? && ::API::V3::Utilities::PathHelper::ApiV3Path::same_origin?(redirect_uri)
-        ServiceResult.success(result: redirect_uri)
-      else
-        ServiceResult.failure
-      end
+      process(@cookies, @state)
+        .fmap { |uri| ServiceResult.success(result: uri) }
+        .value_or ServiceResult.failure
     end
 
     private
 
-    def oauth_state_cookie
-      return nil if @state.blank?
+    def process(cookies, state)
+      uuid = yield state
```

Comment:
I find the name `uuid` is taking me offtrack. It could be almost any string and it doesn't have to be a UUID. For me that is a `state_key`.

```suggestion
      state_key = yield state
```